designs/helm-charts/local-bind-volumes.md

```diff
@@ -0,0 +1,119 @@
+     - Mayastor provisions the volume on the same node as the pod, enforcing local attachment.
+   - If `volumeBindingMode: Immediate` in the storage class for Mayastor:
+     - Mayastor provisions the volume with the `local` flag enabled.
+     - Kubernetes scheduler will schedule the pod on the node where the volume is provisioned, ensuring local attachment.
+
+2. **Single-Replica Restriction**:
+   - This local feature only applies to volumes with a single replica. If additional replicas are added, the local flag becomes ineffective, and the volume no longer maintains strict local binding.
+
+3. **Anti-Rescheduling Enforcement**:
+   - Node affinity rules are set to ensure that a pod cannot be rescheduled to a node without the volume.
+   - If the node with the local volume goes offline, the pod remains unscheduled rather than migrating, preventing data access issues due to cross-node scheduling.
+
+4. **Configuring the Local Flag**:
+   - The `local` flag can be modified by administrators using `kubectl-mayastor` commands, allowing on-demand toggling.
+   - Setting `local: true` enables strict local attachment; setting `local: false` removes the affinity restrictions, allowing the volume to be scheduled across nodes if required.
+
+## 6. User Stories
+
+1. **Story 1**: As an application developer, I want my application pod and its Mayastor volume to stay on the same node, ensuring low latency and high performance for data access.
+2. **Story 2**: As a DevOps engineer, I want control over the local behavior for volumes, so I can enable or disable this feature as needed, using `kubectl-mayastor`.
+3. **Story 3**: As a system administrator, I want the flexibility to toggle local binding without disrupting active volumes, providing seamless adaptability for different workload requirements. Additionally, when cordoning a node for maintenance or reboot, I want to ensure that the pods using local volumes will not be rescheduled to other nodes, ensuring the system's stability during such operations.
+
+## 7. Implementation Details
+
+### Design
+
+- **Local Flag Parameter**: 
+  - Add a `local` flag to the volume specification. When `local: true` is set, Mayastor enforces both local provisioning and anti-rescheduling of single-replica volumes.
+  - `kubectl-mayastor` commands can be used to set or unset the `local` flag for active volumes, dynamically controlling local binding.
+
+- **Single-Replica Volume Constraint**: 
+  - This feature applies only to volumes configured with one replica, as multi-replica configurations may require cross-node replication.
+  - If additional replicas are added, the `local` flag is automatically disabled, and cross-node rescheduling may occur.
+
+- **Affinity Management**:
+  - Node affinity rules are applied to the volume based on the `local` flag setting, ensuring that the pod and volume remain co-located.
+  - Kubernetes will use these affinity rules to restrict rescheduling to the initial node only.
```

Comment:
Do these rules help with `Immediate` binding mode? Also can they be modified after initial creation? (since local flag is intended to be modifiable via plugin)

Reply:
local flag is intended to be set in storageclass, and also be modified via plugin.
If volume is created with Immediate and local, then volume can be created on random node and set with proper affinity. K8S scheduler will then schedule pod on specific node.
```
Node Affinity:
  Required Terms:
    Term 0:        openebs.io/csi-node in [mayastor]
```
```
Node Affinity:
  Required Terms:
    Term 0:        openebs.io/nodename=node1
```

Reply:
I see, it kinda reverses the initial step, so basically:
`WaitForFirsConsumer`: brings storage to node where pod is first scheduled on
`Immediate`: brings pod to the node where mayastor chooses to place the replica

Reply:
yes, and what is the confusion or reverse?

Reply:
It's clear for me now, just wanted to highlight this difference, which I think is important.